File: framework/test2/nest/nestedint/nestedint.go
package main

import (
    . "missingjs.com/soda/leetcode"
    "missingjs.com/soda/unittest"
    "missingjs.com/soda/util"
)

var logger = util.Logger
var _used ListNode

func flatNested(niList []*NestedInteger) []int {
    res := make([]int, 0)
    iter := Constructor(niList)
    for iter.HasNext() {
        res = append(res, iter.Next())
    }
    return res
}

type Node struct {
    List  []*NestedInteger
    Index int
}

func (n *Node) IsEnd() bool {
    return n.Index >= len(n.List)
}

func (n *Node) Value() int {
    return n.Current().GetInteger()
}

func (n *Node) Current() *NestedInteger {
    return n.List[n.Index]
}

type Stack []*Node

func (s *Stack) Push(node *Node) {
    (*s) = append(*s, node)
}

func (s *Stack) Pop() *Node {
    top := s.Top()
    (*s) = (*s)[:s.Size()-1]
    return top
}

func (s *Stack) Size() int {
    return len(*s)
}

func (s *Stack) Top() *Node {
    return (*s)[s.Size()-1]
}

type NestedIterator struct {
    Stack Stack
}

func Constructor(nestedList []*NestedInteger) *NestedIterator {
    iter := &NestedIterator{}
    iter.Stack.Push(&Node{nestedList, 0})
    iter.locate()
    return iter
}

func (this *NestedIterator) locate() {
    for this.Stack.Size() > 0 {
        if this.Stack.Top().IsEnd() {
            this.Stack.Pop()
            if this.Stack.Size() > 0 {
                this.Stack.Top().Index++
            }
        } else if this.Stack.Top().Current().IsInteger() {
            break
        } else {
            this.Stack.Push(&Node{this.Stack.Top().Current().GetList(),0})
        }
    }
}

func (this *NestedIterator) Next() int {
    value := this.Stack.Top().Value()
    this.Stack.Top().Index++
    this.locate()
    return value
}

func (this *NestedIterator) HasNext() bool {
    return this.Stack.Size() > 0 && !this.Stack.Top().IsEnd()
}

func main() {
    // create tester by function
    work := unittest.CreateWork(flatNested)
    // OR create by struct factory
    // work := unittest.CreateWorkForStruct(Constructor)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.Run()
}

File: framework/test2/nest/0341/Nestedint.scala
import soda.scala.unittest._

import scala.reflect.runtime.universe.typeOf

import soda.scala.leetcode._

import collection.mutable
import util.control.Breaks.{break, breakable}
object Solution {
  def flatNested(niList: List[NestedInteger]): List[Int] = {
    val res = mutable.ArrayBuffer[Int]()
    val iter = new NestedIterator(niList)
    while (iter.hasNext()) {
      res.append(iter.next())
    }
    res.toList
  }
}

class NestedIterator(_nestedList: List[NestedInteger]) {

  class Node(list: List[NestedInteger]) {
    var index: Int = 0
    def isEnd: Boolean = index >= list.size
    def value: Int = current.getInteger
    def current: NestedInteger = list(index)
  }

  val stk = mutable.ArrayDeque[Node]()

  stk.append(new Node(_nestedList))
  locate()

  private def locate(): Unit = {
    breakable {
      while (stk.nonEmpty) {
        if (stk.last.isEnd) {
          stk.removeLast()
          if (stk.nonEmpty) {
            stk.last.index += 1
          }
        } else if (stk.last.current.isInteger) {
          break()
        } else {
          stk.append(new Node(stk.last.current.getList.toList))
        }
      }
    }
  }

  def next(): Int = {
    val value = stk.last.value
    stk.last.index += 1
    locate()
    value
  }
  
  def hasNext(): Boolean = {
    stk.nonEmpty && !stk.last.isEnd
  }
}

class Solution {}

class Nestedint {
  def get(): TestWork = {
    val work = TestWork.forObject(typeOf[Solution], "flatNested")
    // val work = TestWork.forStruct(...)
    // work.setValidator((R, R) => Boolean)
    work.compareSerial = true
    work
  }
}

object Nestedint extends App {
  new Nestedint().get().run()
}
